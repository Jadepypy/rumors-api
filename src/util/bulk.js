const BATCH_SIZE = 100;

export default class Bulk {
  constructor(client, batchSize = BATCH_SIZE) {
    this.client = client;
    this.batchSize = batchSize;
    this._operations = [];
    this.actionsCount = 0;
  }

  async push(items, count = 1) {
    this._operations.push(...items);
    this.actionsCount += count;
    if (
      this.actionCounts > this.batchSize ||
      this._operations.length > 3 * this.batchSize
    ) {
      await this.flush();
    }
    return this;
  }

  async flush() {
    if (this._operations.length === 0) {
      return this;
    }

    // Process entires queued up
    const op = this._operations.splice(0);
    this.actionCounts = 0;
    const { body } = await this.client.bulk({
      body: op,
      refresh: 'true',
    });
    if (body.errors) {
      console.error(body.errors);
    } else {
      console.info(`${body.items.length} items processed`);
    }

    return this;
  }
}
