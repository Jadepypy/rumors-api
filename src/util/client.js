import elasticsearch from '@elastic/elasticsearch';

export default new elasticsearch.Client({
  node: process.env.ELASTICSEARCH_URL,
});

// Processes {_id, _version, found, _source: {...}} to
// {id, ..._source}.
//
export function processMeta({
  _id: id,
  _source: source,

  found, // for mget queries
  _score, // for search queries

  sort, // cursor when sorted
}) {
  if (found || _score !== undefined) {
    return { id, ...source, _cursor: sort, _score };
  }
  return null; // not found
}
