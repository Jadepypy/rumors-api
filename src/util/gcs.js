import { Storage } from '@google-cloud/storage';

const storage = new Storage({
  credentials: JSON.parse(process.env.GCS_CREDENTIALS || '{}'),
});
const gcsBucket = storage.bucket(process.env.GCS_BUCKET_NAME || 'default');

/**
 * Generate hash for identifying if two files are similar
 *
 * @param {ReadableStream} fileStream
 * @param {string} fileName
 * @param {string} contentType MIME type
 * @returns {string} url
 */
export async function uploadToGCS(fileStream, fileName, contentType) {
  if (!process.env.GCS_BUCKET_NAME) {
    throw new Error('GCS_BUCKET_NAME is not set, cannot upload file.');
  }

  const options = {
    metadata: {
      contentType,
    },
  };

  const file = gcsBucket.file(fileName);
  const url = await new Promise((resolve, reject) => {
    fileStream
      .pipe(file.createWriteStream(options))
      .on('error', function(err) {
        reject(new Error(`[GCS] ${err}`));
      })
      .on('finish', function() {
        // The file upload is complete.
        resolve(
          `https://storage.googleapis.com/${
            process.env.GCS_BUCKET_NAME
          }/${fileName}`
        );
      });
  });

  return url;
}
