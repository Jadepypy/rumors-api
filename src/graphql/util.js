import {
  GraphQLInputObjectType,
  GraphQLObjectType,
  GraphQLString,
  GraphQLInt,
  GraphQLList,
  GraphQLEnumType,
  GraphQLFloat,
  GraphQLNonNull,
  GraphQLID,
  GraphQLBoolean,
} from 'graphql';

import Connection from './interfaces/Connection';
import Edge from './interfaces/Edge';
import PageInfo from './interfaces/PageInfo';
import Highlights from './models/Highlights';
import client from 'util/client';
import { imageHash } from 'image-hash';

// https://www.graph.cool/docs/tutorials/designing-powerful-apis-with-graphql-query-parameters-aing7uech3
//
// Filtering args definition & parsing
//

/**
 * @param {string} typeName
 * @param {GraphQLScalarType} argType
 * @param {string} description
 * @returns {GraphQLInputObjectType}
 */
function getArithmeticExpressionType(typeName, argType, description) {
  return new GraphQLInputObjectType({
    name: typeName,
    description,
    fields: {
      LT: { type: argType },
      LTE: { type: argType },
      GT: { type: argType },
      GTE: { type: argType },
      EQ: { type: argType },
    },
  });
}

export const timeRangeInput = getArithmeticExpressionType(
  'TimeRangeInput',
  GraphQLString,
  'List only the entries that were created between the specific time range. ' +
    'The time range value is in elasticsearch date format (https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-date-format.html)'
);
export const intRangeInput = getArithmeticExpressionType(
  'RangeInput',
  GraphQLInt,
  'List only the entries whose field match the criteria.'
);

/**
 * @param {object} arithmeticFilterObj - {LT, LTE, GT, GTE, EQ}, the structure returned by getArithmeticExpressionType
 * @returns {object} Elasticsearch range filter param
 * @see https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-range-query.html#range-query-field-params
 */
export function getRangeFieldParamFromArithmeticExpression(
  arithmeticFilterObj
) {
  // EQ overrides all other operators
  if (typeof arithmeticFilterObj.EQ !== 'undefined') {
    return {
      gte: arithmeticFilterObj.EQ,
      lte: arithmeticFilterObj.EQ,
    };
  }

  const conditionEntries = Object.entries(arithmeticFilterObj);

  if (conditionEntries.length === 0) throw new Error('Invalid Expression!');

  return Object.fromEntries(
    conditionEntries.map(([key, value]) => [key.toLowerCase(), value])
  );
}

export const moreLikeThisInput = new GraphQLInputObjectType({
  name: 'MoreLikeThisInput',
  description:
    'Parameters for Elasticsearch more_like_this query.\n' +
    'See: https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-mlt-query.html',
  fields: {
    like: {
      type: GraphQLString,
      description: 'The text string to search for.',
    },
    minimumShouldMatch: {
      type: GraphQLString,
      description:
        'more_like_this query\'s "minimum_should_match" query param.\n' +
        'See https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-minimum-should-match.html for possible values.',
    },
  },
});

export function createFilterType(typeName, args) {
  const filterType = new GraphQLInputObjectType({
    name: typeName,
    fields: () => ({
      ...args,
      // TODO: converting nested AND / OR to elasticsearch
      // AND: { type: new GraphQLList(filterType) },
      // OR: { type: new GraphQLList(filterType) },
    }),
  });
  return filterType;
}

//
// Sort args definition & parsing
//

const SortOrderEnum = new GraphQLEnumType({
  name: 'SortOrderEnum',
  values: {
    ASC: { value: 'asc' },
    DESC: { value: 'desc' },
  },
});

/**
 * @param {string} typeName
 * @param {Array<string|{name: string, description: string}>} filterableFields
 * @returns {GraphQLList<GarphQLInputObjectType>} sort input type for an field input argument.
 */
export function createSortType(typeName, filterableFields = []) {
  return new GraphQLList(
    new GraphQLInputObjectType({
      name: typeName,
      description:
        'An entry of orderBy argument. Specifies field name and the sort order. Only one field name is allowd per entry.',
      fields: filterableFields.reduce((fields, field) => {
        const fieldName = typeof field === 'string' ? field : field.name;
        const description =
          typeof field === 'string' ? undefined : field.description;

        return {
          ...fields,
          [fieldName]: { type: SortOrderEnum, description },
        };
      }, {}),
    })
  );
}

export const pagingArgs = {
  first: {
    type: GraphQLInt,
    description: 'Returns only first <first> results',
    defaultValue: 10,
  },
  after: {
    type: GraphQLString,
    description:
      'Specify a cursor, returns results after this cursor. cannot be used with "before".',
  },
  before: {
    type: GraphQLString,
    description:
      'Specify a cursor, returns results before this cursor. cannot be used with "after".',
  },
};

/**
 * @param {object[]} orderBy - sort input object type
 * @param {{[string]: (order: object) => object}} fieldFnMap - Defines one elasticsearch sort argument entry for a field
 * @returns {Array<{[string]: {order: string}}>} Elasticsearch sort argument in query body
 */
export function getSortArgs(orderBy, fieldFnMap = {}) {
  return orderBy
    .map(item => {
      const field = Object.keys(item)[0];
      const order = item[field];
      const defaultFieldFn = o => ({ [field]: { order: o } });

      return (fieldFnMap[field] || defaultFieldFn)(order);
    })
    .concat({ _id: { order: 'desc' } }); // enforce at least 1 sort order for pagination
}

// sort: [{fieldName: {order: 'desc'}}, {fieldName2: {order: 'desc'}}, ...]
// This utility function reverts the direction of each sort params.
//
function reverseSortArgs(sort) {
  if (!sort) return undefined;
  return sort.map(item => {
    const field = Object.keys(item)[0];
    const order = item[field].order === 'desc' ? 'asc' : 'desc';
    return {
      [field]: {
        ...item[field],
        order,
      },
    };
  });
}

// Export for custom resolveEdges() and resolveLastCursor()
//
export function getCursor(cursor) {
  return Buffer.from(JSON.stringify(cursor)).toString('base64');
}

export function getSearchAfterFromCursor(cursor) {
  if (!cursor) return undefined;
  return JSON.parse(Buffer.from(cursor, 'base64').toString('utf8'));
}

async function defaultResolveTotalCount({
  first, // eslint-disable-line no-unused-vars
  before, // eslint-disable-line no-unused-vars
  after, // eslint-disable-line no-unused-vars
  ...searchContext
}) {
  return (await client.count({
    ...searchContext,
    body: {
      ...searchContext.body,

      // totalCount cannot support these
      sort: undefined,
      track_scores: undefined,
    },
  })).body.count;
}

export async function defaultResolveEdges(
  { first, before, after, ...searchContext },
  args,
  { loaders }
) {
  if (before && after) {
    throw new Error('Use of before & after is prohibited.');
  }

  const nodes = await loaders.searchResultLoader.load({
    ...searchContext,
    body: {
      ...searchContext.body,
      size: first,
      search_after: getSearchAfterFromCursor(before || after),

      // if "before" is given, reverse the sort order and later reverse back
      //
      sort: before
        ? reverseSortArgs(searchContext.body.sort)
        : searchContext.body.sort,
      highlight: {
        order: 'score',
        fields: {
          text: {
            number_of_fragments: 1, // Return only 1 piece highlight text
            fragment_size: 200, // word count of highlighted fragment
            type: 'plain',
          },
          reference: {
            number_of_fragments: 1, // Return only 1 piece highlight text
            fragment_size: 200, // word count of highlighted fragment
            type: 'plain',
          },
        },
        pre_tags: ['<HIGHLIGHT>'],
        post_tags: ['</HIGHLIGHT>'],
      },
    },
  });

  if (before) {
    nodes.reverse();
  }

  return nodes.map(
    ({ _score: score, highlight, inner_hits, _cursor, ...node }) => ({
      node,
      cursor: getCursor(_cursor),
      score,
      highlight,
      inner_hits,
    })
  );
}

async function defaultResolveLastCursor(
  {
    first, // eslint-disable-line no-unused-vars
    before, // eslint-disable-line no-unused-vars
    after, // eslint-disable-line no-unused-vars
    ...searchContext
  },
  args,
  { loaders }
) {
  const lastNode = (await loaders.searchResultLoader.load({
    ...searchContext,
    body: {
      ...searchContext.body,
      sort: reverseSortArgs(searchContext.body.sort),
    },
    size: 1,
  }))[0];

  return lastNode && getCursor(lastNode._cursor);
}

async function defaultResolveFirstCursor(
  {
    first, // eslint-disable-line no-unused-vars
    before, // eslint-disable-line no-unused-vars
    after, // eslint-disable-line no-unused-vars
    ...searchContext
  },
  args,
  { loaders }
) {
  const firstNode = (await loaders.searchResultLoader.load({
    ...searchContext,
    size: 1,
  }))[0];

  return firstNode && getCursor(firstNode._cursor);
}

async function defaultResolveHighlights(edge) {
  const { highlight: { text, reference } = {}, inner_hits } = edge;

  const hyperlinks = inner_hits?.hyperlinks.hits.hits?.map(
    ({
      _source: { url },
      highlight: {
        'hyperlinks.title': title,
        'hyperlinks.summary': summary,
      } = {},
    }) => ({
      url,
      title: title ? title[0] : undefined,
      summary: summary ? summary[0] : undefined,
    })
  );

  // Elasticsearch highlight returns an array because it can be multiple fragments,
  // We directly returns first element(text, title, summary) here because we set number_of_fragments to 1.
  return {
    text: text ? text[0] : undefined,
    reference: reference ? reference[0] : undefined,
    hyperlinks,
  };
}

// All search
//
export function createConnectionType(
  typeName,
  nodeType,
  {
    // Default resolvers
    resolveTotalCount = defaultResolveTotalCount,
    resolveEdges = defaultResolveEdges,
    resolveLastCursor = defaultResolveLastCursor,
    resolveFirstCursor = defaultResolveFirstCursor,
    resolveHighlights = defaultResolveHighlights,
  } = {}
) {
  return new GraphQLObjectType({
    name: typeName,
    interfaces: [Connection],
    fields: () => ({
      totalCount: {
        type: new GraphQLNonNull(GraphQLInt),
        description:
          'The total count of the entire collection, regardless of "before", "after".',
        resolve: resolveTotalCount,
      },
      edges: {
        type: new GraphQLNonNull(
          new GraphQLList(
            new GraphQLNonNull(
              new GraphQLObjectType({
                name: `${typeName}Edge`,
                interfaces: [Edge],
                fields: {
                  node: { type: new GraphQLNonNull(nodeType) },
                  cursor: { type: new GraphQLNonNull(GraphQLString) },
                  score: { type: GraphQLFloat },
                  highlight: {
                    type: Highlights,
                    resolve: resolveHighlights,
                  },
                },
              })
            )
          )
        ),
        resolve: resolveEdges,
      },
      pageInfo: {
        type: new GraphQLNonNull(
          new GraphQLObjectType({
            name: `${typeName}PageInfo`,
            interfaces: [PageInfo],
            fields: {
              lastCursor: {
                type: GraphQLString,
                resolve: resolveLastCursor,
              },
              firstCursor: {
                type: GraphQLString,
                resolve: resolveFirstCursor,
              },
            },
          })
        ),
        resolve: params => params,
      },
    }),
  });
}

/**
 * @param {{status: T}[]} entriesWithStatus - list of objects with "status" field
 * @param {T[]} statuses - list of status to keep
 * @returns {Object[]}
 */
export function filterByStatuses(entriesWithStatus, statuses) {
  return entriesWithStatus.filter(({ status }) => statuses.includes(status));
}

/**
 * @param {Buffer} fileBuffer
 * @param {ArticleTypeEnum} type The article type
 * @returns {string} The hash for identifying if two files are similar
 */
export async function getMediaFileHash(fileBuffer, type) {
  let hash = '';
  if (type === 'IMAGE') {
    hash = await new Promise((resolve, reject) => {
      imageHash({ data: fileBuffer }, 16, true, (error, data) => {
        if (error) {
          console.error(error);
          reject(error);
        } else {
          resolve(data);
        }
      });
    });
  }

  return hash;
}

export const DEFAULT_ARTICLE_REPLY_STATUSES = ['NORMAL'];
export const DEFAULT_ARTICLE_CATEGORY_STATUSES = ['NORMAL'];
export const DEFAULT_REPLY_REQUEST_STATUSES = ['NORMAL'];
export const DEFAULT_ARTICLE_REPLY_FEEDBACK_STATUSES = ['NORMAL'];

/**
 * @param {string} pluralEntityName - the name to display on argument description
 * @returns {object} GraphQL args for common list filters
 */
export function createCommonListFilter(pluralEntityName) {
  return {
    appId: {
      type: GraphQLString,
      description: `Show only ${pluralEntityName} created by a specific app.`,
    },
    userId: {
      type: GraphQLString,
      description: `Show only ${pluralEntityName} created by the specific user.`,
    },
    createdAt: {
      type: timeRangeInput,
      description: `List only the ${pluralEntityName} that were created between the specific time range.`,
    },
    ids: {
      type: new GraphQLList(new GraphQLNonNull(GraphQLID)),
      description: `If given, only list out ${pluralEntityName} with specific IDs`,
    },
    selfOnly: {
      type: GraphQLBoolean,
      description: `Only list the ${pluralEntityName} created by the currently logged in user`,
    },
  };
}

/**
 * Attach (mutates) filterQueries with Elasticsearch query objects by args.filter in GraphQL resolver
 *
 * @param {Array<Object>} filterQueries - list of filter queries of Elasticsearch bool query
 * @param {object} filter - args.filter in resolver
 * @param {string} userId - userId for the currently logged in user
 * @param {string} appid - appId for the currently logged in user
 */
export function attachCommonListFilter(filterQueries, filter, userId, appId) {
  ['userId', 'appId'].forEach(field => {
    if (!filter[field]) return;
    filterQueries.push({ term: { [field]: filter[field] } });
  });

  if (filter.createdAt) {
    filterQueries.push({
      range: {
        createdAt: getRangeFieldParamFromArithmeticExpression(filter.createdAt),
      },
    });
  }

  if (filter.ids) {
    filterQueries.push({ ids: { values: filter.ids } });
  }

  if (filter.selfOnly) {
    if (!userId) throw new Error('selfOnly can be set only after log in');
    filterQueries.push({ term: { userId } }, { term: { appId } });
  }
}
