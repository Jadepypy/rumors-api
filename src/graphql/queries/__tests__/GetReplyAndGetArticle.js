import gql from 'util/GraphQL';
import { loadFixtures, unloadFixtures } from 'util/fixtures';
import fixtures from '../__fixtures__/GetReplyAndArticle';

describe('GetReplyAndGetArticle', () => {
  beforeAll(() => loadFixtures(fixtures));

  describe('GetArticle', () => {
    it('should get the specified article & associated replies from ID', async () => {
      expect(
        await gql`
          {
            GetArticle(id: "foo") {
              text
              references {
                type
              }
              replyCount
              articleReplies(status: NORMAL) {
                status
                canUpdateStatus
                createdAt
                updatedAt
                articleId
                article {
                  id
                }
                replyId
                reply {
                  id
                  text
                  type
                  reference
                }
              }
              replyRequestCount
              replyRequests {
                reason
                feedbackCount
                positiveFeedbackCount
                negativeFeedbackCount
                article {
                  id
                }
              }
              requestedForReply
            }
          }
        `({}, { user: { id: 'fakeUser', appId: 'LINE' } })
      ).toMatchSnapshot();
    });

    it('fetches more than 10 reply requests', async () => {
      expect(
        await gql`
          {
            GetArticle(id: "manyRequests") {
              text
              replyRequestCount
              replyRequests {
                reason
              }
            }
          }
        `({}, { userId: 'fakeUser', appId: 'LINE' })
      ).toMatchSnapshot();
    });

    it('should allow filtering article replies', async () => {
      expect(
        await gql`
          {
            GetArticle(id: "foo3") {
              articleReplies {
                replyId
                reply {
                  id
                }
                status
              }
              normalReplies: articleReplies(status: NORMAL) {
                replyId
                reply {
                  id
                }
                status
              }
              deletedReplies: articleReplies(status: DELETED) {
                replyId
                reply {
                  id
                }
                status
              }
            }
          }
        `({}, { userId: 'fakeUser', appId: 'LINE' })
      ).toMatchSnapshot();

      // Test articleReplies's common filters
      expect(
        await gql`
          {
            GetArticle(id: "foo") {
              user1Replies: articleReplies(
                statuses: [NORMAL, DELETED]
                userId: "user1"
              ) {
                replyId
                userId
                appId
              }
              app1Replies: articleReplies(
                statuses: [NORMAL, DELETED]
                appId: "app1"
              ) {
                replyId
                userId
                appId
              }
              selfOnlyReplies: articleReplies(
                statuses: [NORMAL, DELETED]
                selfOnly: true
              ) {
                replyId
                userId
                appId
              }
            }
          }
        `({}, { userId: 'user2', appId: 'app2' })
      ).toMatchInlineSnapshot(`
        Object {
          "data": Object {
            "GetArticle": Object {
              "app1Replies": Array [
                Object {
                  "appId": "app1",
                  "replyId": "bar2",
                  "userId": "user2",
                },
                Object {
                  "appId": "app1",
                  "replyId": "bar",
                  "userId": "user1",
                },
              ],
              "selfOnlyReplies": Array [
                Object {
                  "appId": "app2",
                  "replyId": "bar3",
                  "userId": "user2",
                },
              ],
              "user1Replies": Array [
                Object {
                  "appId": "app2",
                  "replyId": "bar4",
                  "userId": "user1",
                },
                Object {
                  "appId": "app1",
                  "replyId": "bar",
                  "userId": "user1",
                },
              ],
            },
          },
        }
      `);
    });

    it('should return empty articleReply when there is none', async () => {
      expect(
        await gql`
          {
            GetArticle(id: "foo2") {
              articleReplies(status: DELETED) {
                replyId
                status
              }
            }
          }
        `({}, { userId: 'fakeUser', appId: 'LINE' })
      ).toMatchSnapshot();
    });

    it('relatedArticles should work', async () => {
      // No param
      //
      expect(
        await gql`
          {
            GetArticle(id: "foo") {
              relatedArticles {
                edges {
                  cursor
                  node {
                    id
                  }
                  highlight {
                    text
                    reference
                    hyperlinks {
                      title
                      summary
                    }
                  }
                  score
                }
              }
            }
          }
        `()
      ).toMatchSnapshot('relatedArticle no-param test');

      // filter
      //
      expect(
        await gql`
          {
            GetArticle(id: "foo") {
              relatedArticles(filter: { replyCount: { GT: 0 } }) {
                edges {
                  cursor
                  node {
                    id
                    text
                  }
                  score
                }
              }
            }
          }
        `()
      ).toMatchSnapshot('relatedArticle filter test');

      // sort
      //
      expect(
        await gql`
          {
            GetArticle(id: "foo") {
              relatedArticles(orderBy: [{ _score: ASC }]) {
                edges {
                  cursor
                  node {
                    id
                    text
                  }
                  score
                }
              }
            }
          }
        `()
      ).toMatchSnapshot('relatedArticle sorting test');
    });

    it('feedbacks should work', async () => {
      expect(
        await gql`
          {
            GetArticle(id: "foo") {
              articleReplies {
                feedbacks {
                  id
                  vote
                }
                positiveFeedbackCount
                negativeFeedbackCount
                feedbackCount
                createdAt
              }
            }
          }
        `()
      ).toMatchSnapshot('feedback loading test');
    });

    it('get specified article and articleCategories with NORMAL status', async () => {
      expect(
        await gql`
          {
            GetArticle(id: "foo") {
              text
              categoryCount
              articleCategories {
                id
                categoryId
                category {
                  id
                  title
                  description
                }
              }
            }
          }
        `()
      ).toMatchSnapshot();
    });

    it('get specified article and articleCategories with DELETED status', async () => {
      expect(
        await gql`
          {
            GetArticle(id: "foo") {
              text
              articleCategories(status: DELETED) {
                categoryId
                category {
                  id
                  title
                  description
                }
              }
            }
          }
        `()
      ).toMatchSnapshot();
    });

    it('authenticated fields returns null for non-logged users', async () => {
      expect(
        await gql`
          {
            GetArticle(id: "foo") {
              requestedForReply
            }
          }
        `({}, { appId: 'LINE' })
      ).toMatchSnapshot();
    });
  });

  describe('GetReply', () => {
    it('should get the specified reply & associated articles from ID', async () => {
      expect(
        await gql`
          {
            GetReply(id: "bar") {
              text
              type
              reference
              articleReplies {
                canUpdateStatus
                articleId
                article {
                  text
                }
              }
            }
          }
        `()
      ).toMatchSnapshot();
    });

    it('should allow filtering article replies', async () => {
      expect(
        await gql`
          {
            GetReply(id: "bar2") {
              articleReplies {
                articleId
                article {
                  id
                }
                status
              }
              normalReplies: articleReplies(status: NORMAL) {
                articleId
                article {
                  id
                }
                status
              }
              deletedReplies: articleReplies(status: DELETED) {
                articleId
                article {
                  id
                }
                status
              }
            }
          }
        `()
      ).toMatchSnapshot();
    });
    it('similarReplies should work', async () => {
      // No param
      //
      expect(
        await gql`
          {
            GetReply(id: "bar") {
              similarReplies {
                edges {
                  cursor
                  node {
                    id
                  }
                  score
                  highlight {
                    text
                    reference
                    hyperlinks {
                      title
                      summary
                    }
                  }
                }
              }
            }
          }
        `()
      ).toMatchSnapshot('similarReply no-param test');

      // sort
      //
      expect(
        await gql`
          {
            GetReply(id: "bar") {
              similarReplies(orderBy: [{ _score: ASC }]) {
                edges {
                  cursor
                  node {
                    id
                  }
                  score
                  highlight {
                    text
                    reference
                    hyperlinks {
                      title
                      summary
                    }
                  }
                }
              }
            }
          }
        `()
      ).toMatchSnapshot('similarReply sorting test');
    });
  });

  afterAll(() => unloadFixtures(fixtures));
});
