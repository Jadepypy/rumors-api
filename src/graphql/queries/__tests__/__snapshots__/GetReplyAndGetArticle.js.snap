// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetReplyAndGetArticle GetArticle authenticated fields returns null for non-logged users 1`] = `
Object {
  "data": Object {
    "GetArticle": Object {
      "requestedForReply": null,
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetArticle feedbacks should work: feedback loading test 1`] = `
Object {
  "data": Object {
    "GetArticle": Object {
      "articleReplies": Array [
        Object {
          "createdAt": "2015-01-01T12:10:33Z",
          "feedbackCount": 1,
          "feedbacks": Array [],
          "negativeFeedbackCount": 1,
          "positiveFeedbackCount": 0,
        },
        Object {
          "createdAt": "2015-01-01T12:10:30Z",
          "feedbackCount": 1,
          "feedbacks": Array [
            Object {
              "id": "foo__bar__test-user__test-app",
              "vote": "UPVOTE",
            },
          ],
          "negativeFeedbackCount": 0,
          "positiveFeedbackCount": 1,
        },
        Object {
          "createdAt": "2015-01-01T12:10:32Z",
          "feedbackCount": 0,
          "feedbacks": Array [],
          "negativeFeedbackCount": 0,
          "positiveFeedbackCount": 0,
        },
        Object {
          "createdAt": "2015-01-01T12:10:31Z",
          "feedbackCount": 0,
          "feedbacks": Array [],
          "negativeFeedbackCount": 0,
          "positiveFeedbackCount": 0,
        },
      ],
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetArticle fetches more than 10 reply requests 1`] = `
Object {
  "data": Object {
    "GetArticle": Object {
      "replyRequestCount": 11,
      "replyRequests": Array [
        Object {
          "reason": "Reason 0",
        },
        Object {
          "reason": "Reason 1",
        },
        Object {
          "reason": "Reason 2",
        },
        Object {
          "reason": "Reason 3",
        },
        Object {
          "reason": "Reason 4",
        },
        Object {
          "reason": "Reason 5",
        },
        Object {
          "reason": "Reason 6",
        },
        Object {
          "reason": "Reason 7",
        },
        Object {
          "reason": "Reason 8",
        },
        Object {
          "reason": "Reason 9",
        },
        Object {
          "reason": "Reason 10",
        },
      ],
      "text": "Popular",
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetArticle get specified article and articleCategories with DELETED status 1`] = `
Object {
  "data": Object {
    "GetArticle": Object {
      "articleCategories": Array [
        Object {
          "category": Object {
            "description": "詐騙貼圖、假行銷手法。",
            "id": "c2",
            "title": "免費訊息詐騙",
          },
          "categoryId": "c2",
        },
      ],
      "text": "Lorum ipsum",
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetArticle get specified article and articleCategories with NORMAL status 1`] = `
Object {
  "data": Object {
    "GetArticle": Object {
      "articleCategories": Array [
        Object {
          "category": Object {
            "description": "對同性婚姻的恐懼、愛滋病的誤解與的防疫相關釋疑。",
            "id": "c1",
            "title": "性少數與愛滋病",
          },
          "categoryId": "c1",
          "id": "foo__c1",
        },
      ],
      "text": "Lorum ipsum",
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetArticle relatedArticles should work: relatedArticle filter test 1`] = `
Object {
  "data": Object {
    "GetArticle": Object {
      "relatedArticles": Object {
        "edges": Array [
          Object {
            "cursor": "WyJmb28zIl0=",
            "node": Object {
              "id": "foo3",
              "text": "Lorum ipsum Lorum ipsum Lorum ipsum",
            },
            "score": 0.9489213,
          },
          Object {
            "cursor": "WyJmb28yIl0=",
            "node": Object {
              "id": "foo2",
              "text": "Lorum ipsum Lorum ipsum",
            },
            "score": 0.9210748,
          },
        ],
      },
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetArticle relatedArticles should work: relatedArticle no-param test 1`] = `
Object {
  "data": Object {
    "GetArticle": Object {
      "relatedArticles": Object {
        "edges": Array [
          Object {
            "cursor": "WyJmb28zIl0=",
            "node": Object {
              "id": "foo3",
              "text": "Lorum ipsum Lorum ipsum Lorum ipsum",
            },
            "score": 0.9489213,
          },
          Object {
            "cursor": "WyJmb28yIl0=",
            "node": Object {
              "id": "foo2",
              "text": "Lorum ipsum Lorum ipsum",
            },
            "score": 0.9210748,
          },
        ],
      },
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetArticle relatedArticles should work: relatedArticle sorting test 1`] = `
Object {
  "data": Object {
    "GetArticle": Object {
      "relatedArticles": Object {
        "edges": Array [
          Object {
            "cursor": "WzAuOTIxMDc0OCwiZm9vMiJd",
            "node": Object {
              "id": "foo2",
              "text": "Lorum ipsum Lorum ipsum",
            },
            "score": 0.9210748,
          },
          Object {
            "cursor": "WzAuOTQ4OTIxMywiZm9vMyJd",
            "node": Object {
              "id": "foo3",
              "text": "Lorum ipsum Lorum ipsum Lorum ipsum",
            },
            "score": 0.9489213,
          },
        ],
      },
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetArticle should allow filtering article replies 1`] = `
Object {
  "data": Object {
    "GetArticle": Object {
      "articleReplies": Array [
        Object {
          "reply": Object {
            "id": "fofo",
          },
          "replyId": "fofo",
          "status": "NORMAL",
        },
        Object {
          "reply": Object {
            "id": "bar2",
          },
          "replyId": "bar2",
          "status": "DELETED",
        },
      ],
      "deletedReplies": Array [
        Object {
          "reply": Object {
            "id": "bar2",
          },
          "replyId": "bar2",
          "status": "DELETED",
        },
      ],
      "normalReplies": Array [
        Object {
          "reply": Object {
            "id": "fofo",
          },
          "replyId": "fofo",
          "status": "NORMAL",
        },
      ],
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetArticle should get the specified article & associated replies from ID 1`] = `
Object {
  "data": Object {
    "GetArticle": Object {
      "articleReplies": Array [
        Object {
          "article": Object {
            "id": "foo",
          },
          "articleId": "foo",
          "canUpdateStatus": false,
          "createdAt": "2015-01-01T12:10:30Z",
          "reply": Object {
            "id": "bar",
            "reference": "barbar",
            "text": "bar",
            "type": "NOT_ARTICLE",
          },
          "replyId": "bar",
          "status": "NORMAL",
          "updatedAt": "2015-01-02T12:10:30Z",
        },
      ],
      "references": Array [
        Object {
          "type": "LINE",
        },
      ],
      "replyCount": 1,
      "replyRequestCount": 2,
      "replyRequests": Array [
        Object {
          "feedbackCount": 2,
          "negativeFeedbackCount": 1,
          "positiveFeedbackCount": 1,
          "reason": "Reason foo",
        },
        Object {
          "feedbackCount": 0,
          "negativeFeedbackCount": 0,
          "positiveFeedbackCount": 0,
          "reason": null,
        },
      ],
      "requestedForReply": true,
      "text": "Lorum ipsum",
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetArticle should return empty articleReply when there is none 1`] = `
Object {
  "data": Object {
    "GetArticle": Object {
      "articleReplies": Array [],
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetReply should allow filtering article replies 1`] = `
Object {
  "data": Object {
    "GetReply": Object {
      "articleReplies": Array [
        Object {
          "article": Object {
            "id": "foo",
          },
          "articleId": "foo",
          "status": "DELETED",
        },
        Object {
          "article": Object {
            "id": "foo2",
          },
          "articleId": "foo2",
          "status": "NORMAL",
        },
        Object {
          "article": Object {
            "id": "foo3",
          },
          "articleId": "foo3",
          "status": "DELETED",
        },
      ],
      "deletedReplies": Array [
        Object {
          "article": Object {
            "id": "foo",
          },
          "articleId": "foo",
          "status": "DELETED",
        },
        Object {
          "article": Object {
            "id": "foo3",
          },
          "articleId": "foo3",
          "status": "DELETED",
        },
      ],
      "normalReplies": Array [
        Object {
          "article": Object {
            "id": "foo2",
          },
          "articleId": "foo2",
          "status": "NORMAL",
        },
      ],
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetReply should get the specified reply & associated articles from ID 1`] = `
Object {
  "data": Object {
    "GetReply": Object {
      "articleReplies": Array [
        Object {
          "article": Object {
            "text": "Lorum ipsum",
          },
          "articleId": "foo",
          "canUpdateStatus": true,
        },
      ],
      "reference": "barbar",
      "text": "bar",
      "type": "NOT_ARTICLE",
    },
  },
}
`;
